### ISSUE TITLE
Support forking mode v1

### ISSUE THREAD

**Issue by mensfeld** [karafka/karafka#1058]
We could (I think) easily add an alternative to threading that just runs a cluster via forks for parallel processing.

**Comment by KJTsanaktsidis** [comment]
I've been evaluating Karafka for a project and this would be incredibly useful. 

It seems to me there are two ways to do this:

* Easy: Fork some children, and have each child independently connect to kafka and act as a separate instance of the karafka app
* Hard: Have one process who connects to kafka, and distributes work to a pool of worker processes over some kind of local RPC mechanism

The hard one is hard for a few reasons:
* librdkafka doesn't seem to have (AFAICT) a good mechanism for closing its sockets after fork, so you can't fork from the librdkafka process; you need a separate supervisor process
* You have to marshal kafka operations across the RPC boundary in a synchronous way - for example, if the consumer wants to seek the topic, the next message it receives needs to be the seeked one.
* Ideally, you'd want some kind of userspace shared-memory communication as part of the RPC, which means C extensions and fun (although a unix socket or some such might be "good enough")

But, the hard one is better for one key reason: It allows you to have (CPU) parallelism across _virtual_ partitions as well as real ones (which might be a good selling point for that pro subscription!)

Have you given any thought to which approach you might be interested in? We may need CPU parallelism across virtual partitions for our use-case so I'm looking at how we might achieve this.

**Comment by mensfeld** [comment]
> Have you given any thought to which approach you might be interested in?

Both and more. For the hard one, you can use a mix of dRuby + limitations (no MoM and no LRJ) or a delegation of just the processing without any kafka client operations. I did simple benchmarks, and the overhead of dRuby (which is fairly cheap) kills all the benefits around CPU for short-running jobs or drastically limits them.

The easy is as you described also limited but is also easy ;) and this one will be implemented.

There is, however a third one that will be implemented right after this is fixed or mitigated: https://bugs.ruby-lang.org/issues/19288 - to use ractors for parallel data deserialization and (potentially in the future) work execution with command-based queue for commands delegation. When analyzing CPU non-IO bound intense work, I discovered that deserialization is one of the most CPU-expensive operations, especially with bigger batches. Being able to parallelize this would already bring big benefits to the users.

To summarize:
- forking with independent processes will happen for sure in the near future
- ractors for deserialization are pretty much ready in a POC that I did but are waiting for the mentioned bug or improvements around JSON (for others like avro will create proper issues once json is of the table)
- the "hard" way may be not worth it unless:
  - a) presented with a use-case from a Pro customer where it would actually be needed and would actually provide benefits
  - b) benefits would not be reduced by dRuby synchronization
  - c) the easy mode would not be enough
  - d) ractors would not help

**Comment by KJTsanaktsidis** [comment]
Thanks so much for that super detailed insight!

I might add an e) for the “things to do before hard mode” list as well - I was thinking that some kind of ActiveJob-like wrapper over a forked worker pool might work for our cases; essentially, have the consumer class submit a “job” to the pool over a Unix socket and waiting for it to be done inside the karafka thread. So you have N-many worker processes doing cpu work, being waited on by N-many Karafka threads. 

Has the benefit of zero Karafka changes needed and kind of looks like our existing (resque) job processing if you squint. Anyway mostly something I needed to convince myself was possible “just in case”.


Preforking “easy” parallelism will almost certainly be sufficient. Thanks again!

**Comment by mensfeld** [comment]
> I was thinking that some kind of ActiveJob-like wrapper over a forked worker pool might work for our cases; essentially, have the consumer class submit a “job” to the pool over a Unix socket and waiting for it to be done inside the karafka thread. So you have N-many worker processes doing cpu work, being waited on by N-many Karafka threads.

> Has the benefit of zero Karafka changes needed and kind of looks like our existing (resque) job processing if you squint. Anyway mostly something I needed to convince myself was possible “just in case”.

Not entirely true. You need to keep in mind, that Karafka is not "just" a job / work executor. It provides instrumentation as well as retrying, DLQs, pausing and other functionalities. Those need to be taken into consideration. The same with forks management and retries when killed, etc. This is not an easy thing when you diverge from the extremely happy path.

Karafka above all needs to be stable and reliable. Any features that I introduce need to be tested with full integrations, under many scenarios, etc. This means that edge cases need to be handled.

Simple forking just to save memory via copy-on-write brings limited benefits (as separate subscriptions happen, etc) and this is why I did not implement it yet. You can just run 3 processes and aside from not having average memory savings (especially if YJIT is on) it does not differ that much from forking.

The sophisticated solution needs to be extremely reliable, that is why I am still exploring the ractors for deserialization.

Can you share some light on your processing use-case? Can you assess how big part of CPU work is the data deserialization?

Thanks

**Comment by mensfeld** [comment]
@KJTsanaktsidis it would be also amazing if you could provide me with data/repro that could illustrate the benefits of "complex" forking. This can be achieved via druby dispatch but the question is whether the costs outrun the benefits.

**Comment by KJTsanaktsidis** [comment]
> it would be also amazing if you could provide me with data/repro that could illustrate the benefits of "complex" forking. This can be achieved via druby dispatch but the question is whether the costs outrun the benefits.

Sorry - I can't, because I don't think we really have a use-case for "complex" forking yet. More "insurance" - it might make sense if I outline our current situation.

Currently, we have some very silly Kafka topics with 64 partitions, purely because our existing (not-Karafka) Ruby consumers can only process messages one-at-a-time from a single partition. So, we have a very overprovisioned deployment of 64 consumer processes, each of which is handling a single partition, doing mostly I/O work.

Karafka's thread based concurrency would let us consume multiple partitions from the same process, so we could reduce the number of deployments from 64 down to something more reasonable. However, even better than this: Karafka Pro's virtual partitioning would let us _also_ reduce[1] the number of partitions in the topic - we only need to process messages with the same key in series, not the entire partition, and virtual partitions would let us easily achieve intra-partition concurrency respecting this constraint.

So, let's imagine a future where I've gone and made the number of partitions much lower. Everything is great - the consumer has good CPU utilisation, so our k8s team isn't yelling at me about wastage, and our topic has a resonable number of partitions, so our kafka team isn't yelling at me too.

Now, imagine some new requirements cause the consumer to need to do more CPU work. We can't simply scale it out anymore - the maximum CPU parallelism we can get is equal to the number of partitions, because GVL. In this situation, having druby-based concurrency would let us still use more CPUs as needed, without needing to re-create the topic again with more partitions.

Essentially, I want to substitute a very silly amount of real Kafka partitions for Karafka's virtual partitioning system - but real kafka partitions can benefit from CPU parallelism (by running more kafafka processes), whilst virtual partitions can't. It would be nice if there were some "escape hatches" if it turns out future requirements (that I definitely don't currently have) mean the CPU requirements on this consumer become higher in the future.

Hopefully this explains where I'm coming from? The tl;dr is that really preforking karafka will absolutely be fine for us based on current requirements. My sketch of "have our consumer implementation reach out somewhere else to run the work, and block waiting for completion" in my previous comment is really just making sure I don't design us into a corner.

[1] well, we'd have to re-create the topic, but that's by-the-by.

**Comment by mensfeld** [comment]
> Hopefully this explains where I'm coming from?

It does :)

> My sketch of "have our consumer implementation reach out somewhere else to run the work, and block waiting for completion" in my previous comment is really just making sure I don't design us into a corner.

It is doable. A simple POC that I built for CPU intense workloads using a delegated drb flow shows 5x the improvement against a regular flow in Karafka (even with virtual partitions). It is totally doable but will require some complex engineering to make it really reliable.

That said, this or ractors will be the direction where Karafka goes independently from regular forks. It just takes time ;)

**Comment by mensfeld** [comment]
I think I'm going to provide this in 2.4

**Comment by mensfeld** [comment]
@KJTsanaktsidis with the release of 2.3 that is going to happen in January, I will be working on this (still planned for 2.4). Do you still want to catch up on your expectations/visition for this functionality?

**Comment by mensfeld** [comment]
I think we need to support at least a "normal" mode where we just fork and a mode where we can specify works listener strategy. This would allow us to control subscriptions in those processes in a granular way.

**Comment by KJTsanaktsidis** [comment]
Having shaken a few trees internally at Zendesk - I think, for all of our current use-cases, we're not likely to be very CPU bound. And the most CPU intensive part of our consumers are likely to be deserialisation, which I know you were looking at special ways to handle in the future (e.g. ractors) - it's much more constrained than "our consumer will spend lots of cycles" in general.

So I think a simple "normal" forking mode (in order to save memory) is probably all we would require in the short/medium term.

**Comment by mensfeld** [comment]
Great. This is what I also had in mind, just with a small DSL for ability to control which of the subscriptions should run on forks when not matching 1:1 with the parent. Once 2.3 is released I should focus on that.

**Comment by mensfeld** [comment]
Things worth including:

1. Web UI should NOT report prior to cluster starting OR descriptors need to be closed. This also should be true for any open producer listeners that report errors back to kafka, otherwise segfault
2. Web UI now starts tracking when process is post initialization - this may be too early since we should first fork and then move
3. Aside from a regular forking we should check if it would be possible to use pitchfork approach of forking from active and refork after time period
4. Shutdown and errors should propagate
5. Worth considering reforking on crashes
6. (Maybe) we should provide indication in the web UI of both the fact that process is a fork as well as its parent
7. We should for sure report the supervisor process in web UI as well

**Comment by mensfeld** [comment]
todo:
- [ ] Restart crashed forks
- [ ] Make sure orphanes are handled
- [ ] Support forceful shutdown according to the timeout
